handler/src/test/java/io/netty/handler/flow/FlowControlHandlerTest.java

```diff
@@ -0,0 +1,388 @@
+     * transferred to the server.
+     */
+    @Test
+    public void testAutoReadingOn() throws Exception {
+        final CountDownLatch latch = new CountDownLatch(3);
+
+        ChannelInboundHandlerAdapter handler = new ChannelInboundHandlerAdapter() {
+            @Override
+            public void channelRead(ChannelHandlerContext ctx, Object msg) {
+                ReferenceCountUtil.release(msg);
+                // We're turning off auto reading in the hope that no
+                // new messages are being sent but that is not true.
+                ctx.channel().config().setAutoRead(false);
+
+                latch.countDown();
+            }
+        };
+
+        Channel server = newServer(true, handler);
+        Channel client = newClient(server);
+
+        try {
+            client.writeAndFlush(newOneMessage())
+                .syncUninterruptibly();
+
+            // We received three messages even through auto reading
+            // was turned off after we received the first message.
+            assertTrue(latch.await(1L, TimeUnit.SECONDS));
+        } finally {
+            client.close();
+            server.close();
+        }
+    }
+
+    /**
+     * This test demonstrates the default behavior if auto reading
+     * is turned off from the get-go and you're calling read() in
+     * the hope that only one message will be returned.
+     *
+     * NOTE: This test waits for the client to disconnect which is
+     * interpreted as the signal that all {@code byte}s have been
+     * transferred to the server.
+     */
+    @Test
+    public void testAutoReadingOff() throws Exception {
+        final Exchanger<Channel> peerRef = new Exchanger<Channel>();
+        final CountDownLatch latch = new CountDownLatch(3);
+
+        ChannelInboundHandlerAdapter handler = new ChannelInboundHandlerAdapter() {
+            @Override
+            public void channelActive(ChannelHandlerContext ctx) throws Exception {
+                peerRef.exchange(ctx.channel(), 1L, TimeUnit.SECONDS);
+                ctx.fireChannelActive();
+            }
+
+            @Override
+            public void channelRead(ChannelHandlerContext ctx, Object msg) {
+                ReferenceCountUtil.release(msg);
+                latch.countDown();
+            }
+        };
+
+        Channel server = newServer(false, handler);
+        Channel client = newClient(server);
+
+        try {
+            // The client connection on the server side
+            Channel peer = peerRef.exchange(null, 1L, TimeUnit.SECONDS);
+
+            // Write the message
+            client.writeAndFlush(newOneMessage())
+                .syncUninterruptibly();
+
+            // Read the message
+            peer.read();
+
+            // We received all three messages but hoped that only one
+            // message was read because auto reading was off and we
+            // invoked the read() method only once.
+            assertTrue(latch.await(1L, TimeUnit.SECONDS));
```

Comment:
No because this test demonstrates the default behavior without the FlowControlHandler in the pipeline.